wordpress-multisite

null-git opened the pull request:
A Munin plugin to monitor posts, comments and pingbacks from every multisite instance.
It uses multigraphs and also shows the combined number of posts, comments, pingbacks, instances and users.

Most database requests came from the wordpress-mu-plugin which was written by Andre Darafarin and Chris Bair

sumpfralle commented:
Thank you for this new plugin!

I have some questions and suggestions:
* please specify a license
* would it be possible to change "bash" to "sh"?
    * you would need to replace "=="  with "=" or "-eq"
    * `(( FOO += 1 ))` can be replaced with `FOO=$((FOO + 1))`
* would you mind using "$(...)" instead of the old-fashioned backticks?
* maybe you want to consider using a shell function for your repetitive "mysql ... count" statement
* do you think it could be worthwhile to change your configuration (DBHOST, ...) to the style of the core mysql plugin (env.mysqlopts, env.mysqluser, env.mysqlconnection)? This could allow password-less access (e.g. via the "debian-sys-maint" user on debian-based systems) while still allowing your style of "-H${DBHOST}" arguments.

Anyway: it is your code - everything except for the missing license is optional ...

null-git commented:
Thank you for your suggestions. I've added the license and reworked the code.
I couldn't test this plugin with a pure sh but I'm fairly confident that I have removed all the bashisms.
I've also changed the names of the configuration variables but the plugin still can't access the variable from the [mysql*] section because of its [wordpress_*] section name?

sumpfralle commented:
we are almost there :)

I have a few last nitpicks for you:
* the configuration section is not rendered for me, when I run `perldoc wordpress-multisite` - maybe just remove the "item" formatting and simply indent these lines
* the "VERSION" date is not the same as the one at the top
* please change the shebang from `/usr/bin/env sh` to `/bin/sh` - the "env" approach should only be useful for non-shell interpreters IMO
* if you want to run a syntax check of the script you can use `shellcheck -s dash SCRIPT`
** this will report a good amount of places where it would be advisable to double-quote variables (the only exceptions are probably DB_OPTIONS and DB_CONNECTION)
* extra bonus: add [capabilities](http://munin-monitoring.org/wiki/PluginConcise#Magicmarkers) (especially: autoconf)

sumpfralle commented:
@null-git: what do you think regarding my nitpicks?

Anyway: since the new plugin is already in a good state, I would also be willing to merge it as it is.

null-git commented:
Hey. I've fixed the perldoc issue, fixed the version and got rid of most shellcheck warnings.

evn is also appropriate for shells. I.e. FreeBSD has /usr/local/bin/bash, Arch Linux has /usr/bin/bash and Debian Jessie has /bin/bash. While I don't have an example right now for sh, other projects like Minix ( https://github.com/Stichting-MINIX-Research-Foundation/minix/blob/master/build.sh ) also do this.

sumpfralle commented on plugins/other/wordpress-multisite:
```diff
@@ -0,0 +1,131 @@
+		users) QUERY="SELECT COUNT(*) FROM ${DB_PREFIX}users;"
+	esac
+
+	$MYSQL_CMD $DB_CONNECTION $DB_OPTIONS $DB_NAME --column-names=0 -s --execute="$QUERY"
```
MYSQL_CMD and DB_NAME should be quoted in order to avoid whitespace issues.

null-git replied:
done

sumpfralle commented on plugins/other/wordpress-multisite:
```diff
@@ -0,0 +1,131 @@
+else
+	for n in $(wp_get ids); do
+		i=
+		test "$n" -gt "1" && i=${n}_
```
I have problems grasping the relation between `i` and `n`. Could you please add a comment here?

null-git replied:
See next comment

sumpfralle commented on plugins/other/wordpress-multisite:
```diff
@@ -0,0 +1,131 @@
+for n in $(wp_get ids); do
+	i=
+	test "$n" -gt "1" && i=${n}_
+	test "$n" -le "9" && n=0${n}
```
I understand that you are using this manipulation of `n` in order to influence the sort order of the following multigraphs.
In this case I recommend to pick a longer numerical width and pick an easier understable variable name (e.g. `sort_name=$(printf "%06d" "$n")`). Future readers of this code will surely appreciate this detail.

null-git replied:
Used the printf approach but with configurable number length.
Also changed the variable name and moved the $i part to wp_get() and commented it.

sumpfralle commented on plugins/other/wordpress-multisite:
```diff
@@ -0,0 +1,131 @@
+
+MYSQL_CMD=$(which mysql)
+wp_get() {
+	case $1 in
```
Please add `local QUERY` (or at least expand the `case` with a default branch) in order to avoid hard-to-debug situations.

null-git replied:
done

null-git commented:
Never did anything in perl so I just changed it to /bin/sh for now.

Yeah. /bin is a link to /usr/bin in Arch Linux but it's only there so things like this don't break. But there is no reason why sh shouldn't be anywhere else than /bin so I think the env approach is justified even for sh scripts :)

sumpfralle commented:
Regarding the location of /bin/sh: yes, probably this is really the future. Somehow it still feels wrong, that /usr/bin/env should be the only universal path, we can rely on. It was surely not meant like that :)

Anyway: thank you for your time and this shiny new plugin!

steveschnepp commented:
Indeed, I usually don't really like the `/usr/bin/env` in published scripts as it depends on yet another tool to work.

The usual way is to put a simple path, either in `/bin/` for shells & POSIX stuff (such as `awk`), and `/usr/bin/` for anything else (such as `perl`, `python` or `node`).

Let users just change the path if they have a weird setup 😉
